### PR TITLE
Record `invocation.configSource` section in slsa provenance

### PR DESCRIPTION
**Pull request by chuangw6**
# Changes

Related to https://github.com/tektoncd/chains/issues/521.

Prior to this PR, the SLSA provenance that Chains produces did not record
the source information of remote resources.

In this change, we want to record the source information in the provenance
to track where the remote pipeline/task definition came from i.e. git repo,
tekton bundles etc. The source information is available in
PipelineRun/TaskRun `Status.Provenance` field.

If a remote pipeline definition references remote task definitions from other remote places,
`predicate.invocation.configSource` will record the source information of
the remote pipeline definition only, and `predicate.buildConfig.tasks[x].invocation.configSource`
will record the source information for the corresponding remote task definition.

Signed-off-by: Chuang Wang <chuangw@google.com>

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
SLSA provenance records the source information of remote task/pipeline definition.
```

### CODE REVIEW

**Comment by chuangw6** [comment]
/hold

waiting for https://github.com/tektoncd/pipeline/pull/5397

**Comment by lcarva** [comment]
@chuangw6, I remember a comment from @priyawadhwa that if we populate configSource, we shouldn't update buildConfig. Maybe I misunderstood the comment?

**Comment by chuangw6** [comment]
> @chuangw6, I remember a comment from @priyawadhwa that if we populate configSource, we shouldn't update buildConfig. Maybe I misunderstood the comment?

Yeah, I remember that. Talking to @wlynch before, it seems fine to include both sections if we do want to record the remote refs in the provenance.

Thoughts @priyawadhwa @wlynch?

**Comment by priyawadhwa** [comment]
I'd prefer the provenance be as correct as possible based on the spec. If we can correctly set `configSource` we shouldn't need `buildConfig` anymore!

**Comment by chuangw6** [comment]
> I'd prefer the provenance be as correct as possible based on the spec. If we can correctly set `configSource` we shouldn't need `buildConfig` anymore!

That makes sense to me as long as we still record the param values provided by taskrun/pipeline run in the `invocation.parameters` section. So we can
- record `buildConfig` section only for embedded and incluster definition. 
- record `invocation.configSource` section only for remote resources (git, oci, catalog).

But one case I am concerned about is what if the remote resource that user specified in resolver is not retrievable later for some reason. For example, a user specifies git url, commit and pathInRepo correctly in taskrun definition  -> taskrun reconciler fetches the resource -> Chains records it in `invocation.configSource` and leaves `buildConfig` empty -> later the repo that contains the task definition is deleted. In this case, the information recorded in `invocation.configSource` or say the whole provenance will be meaningless. wdyt?

**Comment by chuangw6** [comment]
Seems like the slsa0.2 spec doesn't explicitly say the `configSource` and `buildConfig` shouldn't exist simultaneously. In the [Cloud Build example](https://slsa.dev/provenance/v0.2#:~:text=Cloud%20Build%20config%2Das%2Dcode) listed on the website, both sections are filled with content.

**Comment by priyawadhwa** [comment]
> That makes sense to me as long as we still record the param values provided by taskrun/pipeline run in the invocation.parameters section. 

SGTM.

Re the Cloud Build example, my guess is the example isn't entirely accurate (the Cloud Build RPC example does specify `// Build steps were provided as an argument. No configSource`). The definition for `buildConfig` is:

> Lists the steps in the build. If invocation.configSource is not available, buildConfig can be used to verify information about the build.

and when I added `buildConfig` into the v0.2 spec a few months ago it was definitely intended as an alternative to `configSource`. I don't really see a use case for having both (if a repo is deleted I'd guess you probably just wouldn't trust the artifact at all), but if you feel strongly we can give it a try.

**Comment by chuangw6** [comment]
Thanks for the details @priyawadhwa !
I feel like it might be necessary to always keep `buildConfig` section for at least auditing purpose and also for helping reproduce a build. Although we set `invocation.configSource` correctly for the remote definitions, it's possible that the consumer of the provenance does not have the access to the remote repo especially the gitresolver recently supports fetching files from a private repo in https://github.com/tektoncd/pipeline/pull/5450.

File: pkg/chains/formats/intotoite6/testdata/taskrun2.json [hunk 1]
```diff
@@ -8,6 +8,11 @@
         "serviceAccountName": "default"
     },
     "status": {
+        "annotations": {
+            "resolution.tekton.dev/path": "task.yaml",
+            "resolution.tekton.dev/revision": "123fdf35b4e7b1a56a84b2796aab2827edd65c25",
+            "resolution.tekton.dev/uri": "https://github.com/test/tekton-test.git"
+        },
```
Comment (wlynch) [inline pkg/chains/formats/intotoite6/testdata/taskrun2.json:15]: Fine for now, but highly recommend working towards making these actual fields in the Run status.

**Review by wlynch — CHANGES_REQUESTED** [review]
@chuangw6 would it be possible to stuff all other fields into `environment`?

The way I see it right now is if PipelineSpec is set, we should set buildConfig, if PipelineRef is set, we set configSource. If there are any values that don't fit nicely with that, let's call those out.

I do see value in recording evaluated specs in the provenance (I'd also like to record the pod that was executed!), but it sounds like the current SLSA 0.2 provenance isn't really meant for that at the moment - this would be a good thing to raise at the SLSA WG.

File: pkg/chains/formats/intotoite6/intotoite6.go [hunk 2]
```diff
@@ -128,6 +130,20 @@ func invocation(tr *v1beta1.TaskRun) slsa.ProvenanceInvocation {
...
+
+	// configSource
+	// ------------
+	// TODO (@chuangw6): replace the annotation keys with the constants from the resolution package
+	// https://github.com/tektoncd/pipeline/pull/5397
+	if tr.Status.Annotations != nil {
+		i.ConfigSource = slsa.ConfigSource{
+			URI: tr.Status.Annotations["resolution.tekton.dev/uri"],
+			Digest: slsa.DigestSet{
+				"sha256": tr.Status.Annotations["resolution.tekton.dev/revision"],
```
[the remark this answers is not in the document]
Comment (chuangw6) [reply pkg/chains/formats/intotoite6/intotoite6.go]: Thanks for comment. The algorithm is now set in the individual resolvers and differs from resolver to resolver. Please take a look.

**Review by wlynch — APPROVED** [review]
/lgtm

**Pull request merged**